File: kakao.app/js/ked/view/menu/context.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }}

var context

import kxk from "../../../kxk.js"
let kstr = kxk.kstr
let slash = kxk.slash
let post = kxk.post
let noon = kxk.noon

import belt from "../../edit/tool/belt.js"

import theme from "../../theme/theme.js"

import inputchoice from "./inputchoice.js"


context = (function ()
{
    _k_.extend(context, inputchoice)
    context["menu"] = null
    function context (screen)
    {
        this.screen = screen
    
        this["draw"] = this["draw"].bind(this)
        this["arrange"] = this["arrange"].bind(this)
        context.__super__.constructor.call(this,this.screen,'context')
        context.menu = this
    }

    context.prototype["arrange"] = function ()
    {
        var c, h, ih, iz, w, x, y

        w = this.width + 2 + 1
        c = this.choices.numChoices()
        ih = (this.inputIsActive() ? 2 : 0)
        iz = _k_.max(0,ih - 1)
        h = c + 2 + ih
        x = this.pos[0]
        y = this.pos[1]
        this.input.layout(x + 2,y + 1,w - 4,iz)
        this.choices.layout(x + 1,y + 1 + ih,w - 2,c)
        return this.cells.layout(x,y,w,h)
    }

    context["show"] = function (pos, cb, items)
    {
        return context.menu.show(pos,cb,items)
    }

    context.prototype["show"] = function (pos, cb, items)
    {
        this.pos = pos
        this.cb = cb
        this.items = items
    
        var items

        this.pos[1] -= 1
        items = this.items.map(function (i)
        {
            return ' ' + i
        })
        this.width = belt.widthOfLines(items)
        this.input.set('')
        this.input.hide()
        this.choices.set(items)
        this.choices.select(0)
        this.choices.state.setView([0,0])
        return context.__super__.show.call(this)
    }

    context.prototype["applyChoice"] = function (choice)
    {
        console.log('context applyChoice',choice)
        this.cb(choice)
        return this.hide()
    }

    context.prototype["onWheel"] = function ()
    {}

    context.prototype["draw"] = function ()
    {
        if (this.hidden())
        {
            return
        }
        this.arrange()
        this.drawFrame()
        return this.drawChoices()
    }

    return context
})()

export default context;

File: kakao.app/love/lua/view/menu/context.lua
--[[
     ███████   ███████   ███   ███  █████████  ████████  ███   ███  █████████  
    ███       ███   ███  ████  ███     ███     ███        ███ ███      ███     
    ███       ███   ███  ███ █ ███     ███     ███████     █████       ███     
    ███       ███   ███  ███  ████     ███     ███        ███ ███      ███     
     ███████   ███████   ███   ███     ███     ████████  ███   ███     ███     

    quick and dirty context menu for the fileeditor
--]]

inputchoice = require "view.menu.inputchoice"


local context = class("context", inputchoice)
    


function context:init() 
        inputchoice.init(self, 'context')
        
        self:setColor('bg', theme.context.bg)
        _G.contextMenu = self
        return self
    end

--  ███████   ████████   ████████    ███████   ███   ███   ███████   ████████
-- ███   ███  ███   ███  ███   ███  ███   ███  ████  ███  ███        ███     
-- █████████  ███████    ███████    █████████  ███ █ ███  ███  ████  ███████ 
-- ███   ███  ███   ███  ███   ███  ███   ███  ███  ████  ███   ███  ███     
-- ███   ███  ███   ███  ███   ███  ███   ███  ███   ███   ███████   ████████


function context:arrange() 
        local w = ((self.width + 2) + 1) -- 2 for frame, 1 for right 
        local c = self.choices:numChoices()
        local ih = (function () 
    if self:inputIsActive() then 
    return 2 else 
    return 0
             end
end)()
        local iz = max(0, (ih - 1))
        local h = ((c + 2) + ih)
        local x = self.pos[1]
        local y = self.pos[2]
        print("nummChoices", c, x, y, w, h, ih, iz)
        
        self.input:layout((x + 2), (y + 1), (w - 4), iz)
        self.choices:layout((x + 1), ((y + 1) + ih), (w - 2), c)
        return self.cells:layout(x, y, w, h)
    end

--  0000000  000   000   0000000   000   000
-- 000       000   000  000   000  000 0 000
-- 0000000   000000000  000   000  000000000
--      000  000   000  000   000  000   000
-- 0000000   000   000   0000000   00     00


function context.static.show(pos, cb, items) 
        if _G.contextMenu then 
            return _G.contextMenu:show(pos, cb, items)
        else 
            return print("DAFURGL?")
        end
    end


function context:show(pos, cb, items) 
        self.pos = pos
        self.cb = cb
        self.items = items
        
        -- @pos[1] -= 1
        
        items = self.items:map(function (i) 
    return ' ' .. i
end) -- padding for  
        
        self.width = belt.widthOfLines(items)
        
        self.input:set('')
        self.input:hide()
        self.choices:set(items)
        self.choices:select(1)
        self.choices.state:setView(array(1, 1))
        
        return inputchoice.show(self)
    end

--  0000000   00000000   00000000   000      000   000  
-- 000   000  000   000  000   000  000       000 000   
-- 000000000  00000000   00000000   000        00000    
-- 000   000  000        000        000         000     
-- 000   000  000        000        0000000     000     


function context:applyChoice(choice) 
        -- log 'context applyChoice' choice
        self.cb(choice)
        return self:hide()
    end


function context:onWheel() 
    
    end

-- 0000000    00000000    0000000   000   000
-- 000   000  000   000  000   000  000 0 000
-- 000   000  0000000    000000000  000000000
-- 000   000  000   000  000   000  000   000
-- 0000000    000   000  000   000  00     00


function context:draw() 
        if self:hidden() then return end
        
        self:arrange()
        self:drawFrame()
        self:drawChoices()
        return self:render()
    end


function context:drawFrame() 
        local cy = 0
        
        if self.input:visible() then 
            self.cells:draw_rounded_border(0, 0, -1, 2, {fg = self.color.bg})
            cy = 2
        end
        
        return self.cells:draw_rounded_border(0, cy, -1, -1, {fg = self.color.bg})
    end

return context